test/integration/scheduler/framework_test.go

```diff
@@ -599,6 +637,166 @@ func TestUnreservePlugin(t *testing.T) {
+		pod, err := createPausePod(cs,
+			initPausePod(cs, &pausePodConfig{Name: "test-pod", Namespace: context.ns.Name}))
+		if err != nil {
+			t.Errorf("Error while creating a test pod: %v", err)
+		}
+
+		if test.expectBoundByScheduler || test.expectBoundByPlugin {
+			// bind plugins skiped to bind the pod
+			if err = waitForPodToSchedule(cs, pod); err != nil {
+				t.Errorf("test #%v: Expected the pod to be scheduled. error: %v", i, err)
+				continue
+			}
+			pod, err = cs.CoreV1().Pods(pod.Namespace).Get(pod.Name, metav1.GetOptions{})
+			if err != nil {
+				t.Errorf("can't get pod: %v", err)
+			}
+			if test.expectBoundByScheduler {
+				if pod.Annotations[bindPluginAnnotation] != "" {
+					t.Errorf("test #%v: Expected the pod to be binded by scheduler instead of by bindplugin %s", i, pod.Annotations[bindPluginAnnotation])
+				}
+				if bindPlugin1.numBindCalled != 1 || bindPlugin2.numBindCalled != 1 {
+					t.Errorf("test #%v: Expected each bind plugin to be called once, was called %d and %d times.", i, bindPlugin1.numBindCalled, bindPlugin2.numBindCalled)
+				}
+			} else {
+				if pod.Annotations[bindPluginAnnotation] != test.expectBindPluginName {
+					t.Errorf("test #%v: Expected the pod to be binded by bindplugin %s instead of by bindplugin %s", i, test.expectBindPluginName, pod.Annotations[bindPluginAnnotation])
+				}
+				if bindPlugin1.numBindCalled != 1 {
+					t.Errorf("test #%v: Expected %s to be called once, was called %d times.", i, bindPlugin1.Name(), bindPlugin1.numBindCalled)
+				}
+				if test.expectBindPluginName == bindPlugin1.Name() && bindPlugin2.numBindCalled > 0 {
+					// expect bindplugin1 succeeded to bind the pod and bindplugin2 should not be called.
+					t.Errorf("test #%v: Expected %s not to be called, was called %d times.", i, bindPlugin2.Name(), bindPlugin1.numBindCalled)
+				}
+			}
+			if err = wait.Poll(10*time.Millisecond, 30*time.Second, func() (done bool, err error) {
+				return ptbdPlugin.numPostbindCalled == 1, nil
+			}); err != nil {
+				t.Errorf("test #%v: Expected the postbind plugin to be called once, was called %d times.", i, ptbdPlugin.numPostbindCalled)
+			}
+			if unresPlugin.numUnreserveCalled != 0 {
+				t.Errorf("test #%v: Expected the unreserve plugin not to be called, was called %d times.", i, unresPlugin.numUnreserveCalled)
+			}
+		} else {
+			// bind plugin fails to bind the pod
+			if err = wait.Poll(10*time.Millisecond, 30*time.Second, podSchedulingError(cs, pod.Namespace, pod.Name)); err != nil {
+				t.Errorf("test #%v: Expected a scheduling error, but didn't get it. error: %v", i, err)
+			}
+			if ptbdPlugin.numPostbindCalled > 0 {
+				t.Errorf("test #%v: Didn't expected the postbind plugin to be called %d times.", i, ptbdPlugin.numPostbindCalled)
+			}
+			if err = wait.Poll(10*time.Millisecond, 30*time.Second, func() (done bool, err error) {
+				// The scheduler may have tried to schedule the pod several times, so we can only assert numUnreserveCalled > 0
```

Comment:
Why did The scheduler try to schedule the pod several times and could you explain the reason for retrying?

Reply:
You can see the log "About to try and schedule pod bind-plugin " for many times in https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/79293/pull-kubernetes-integration/1142426772172705796/

```
I0622 14:00:13.249848  110366 scheduling_queue.go:815] About to try and schedule pod bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/test-pod
I0622 14:00:13.249863  110366 scheduler.go:470] Attempting to schedule pod: bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/test-pod
I0622 14:00:13.250201  110366 scheduler_binder.go:256] AssumePodVolumes for pod "bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/test-pod", node "test-node-1"
I0622 14:00:13.250216  110366 scheduler_binder.go:266] AssumePodVolumes for pod "bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/test-pod", node "test-node-1": all PVCs bound and nothing to do
E0622 14:00:13.250264  110366 framework.go:274] bind plugin bind-plugin-1 failed to bind pod bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/test-pod: failed to bind
I0622 14:00:13.250281  110366 scheduler.go:438] Failed to bind pod: bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/test-pod
E0622 14:00:13.250298  110366 factory.go:678] Error scheduling bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/test-pod: scheduler RunBindPlugins failed for pod bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/test-pod: code 1, err <nil>; retrying
I0622 14:00:13.250328  110366 factory.go:736] Updating pod condition for bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/test-pod to (PodScheduled==False, Reason=SchedulerError)
E0622 14:00:13.250343  110366 scheduler.go:596] error binding pod: scheduler RunBindPlugins failed for pod bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/test-pod: code 1, err <nil>
I0622 14:00:13.252746  110366 wrap.go:47] GET /api/v1/namespaces/bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/pods/test-pod: (1.76251ms) 200 [scheduler.test/v0.0.0 (linux/amd64) kubernetes/$Format 127.0.0.1:37400]
I0622 14:00:13.253554  110366 reflector.go:243] k8s.io/client-go/informers/factory.go:133: forcing resync
I0622 14:00:14.152063  110366 wrap.go:47] GET /api/v1/namespaces/default: (1.615006ms) 200 [scheduler.test/v0.0.0 (linux/amd64) kubernetes/$Format 127.0.0.1:37400]
I0622 14:00:14.153778  110366 wrap.go:47] GET /api/v1/namespaces/default/services/kubernetes: (1.326618ms) 200 [scheduler.test/v0.0.0 (linux/amd64) kubernetes/$Format 127.0.0.1:37400]
I0622 14:00:14.155321  110366 wrap.go:47] GET /api/v1/namespaces/default/endpoints/kubernetes: (1.176727ms) 200 [scheduler.test/v0.0.0 (linux/amd64) kubernetes/$Format 127.0.0.1:37400]
I0622 14:00:14.246811  110366 reflector.go:243] k8s.io/client-go/informers/factory.go:133: forcing resync
I0622 14:00:14.246866  110366 reflector.go:243] k8s.io/client-go/informers/factory.go:133: forcing resync
I0622 14:00:14.246880  110366 reflector.go:243] k8s.io/client-go/informers/factory.go:133: forcing resync
I0622 14:00:14.246954  110366 reflector.go:243] k8s.io/client-go/informers/factory.go:133: forcing resync
I0622 14:00:14.249982  110366 reflector.go:243] k8s.io/client-go/informers/factory.go:133: forcing resync
I0622 14:00:14.250109  110366 scheduling_queue.go:815] About to try and schedule pod bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/test-pod
I0622 14:00:14.250123  110366 scheduler.go:470] Attempting to schedule pod: bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/test-pod
I0622 14:00:14.250385  110366 scheduler_binder.go:256] AssumePodVolumes for pod "bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/test-pod", node "test-node-0"
I0622 14:00:14.250412  110366 scheduler_binder.go:266] AssumePodVolumes for pod "bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/test-pod", node "test-node-0": all PVCs bound and nothing to do
E0622 14:00:14.250465  110366 framework.go:274] bind plugin bind-plugin-1 failed to bind pod bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/test-pod: failed to bind
I0622 14:00:14.250484  110366 scheduler.go:438] Failed to bind pod: bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/test-pod
E0622 14:00:14.250502  110366 factory.go:678] Error scheduling bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/test-pod: scheduler RunBindPlugins failed for pod bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/test-pod: code 1, err <nil>; retrying
I0622 14:00:14.250542  110366 factory.go:736] Updating pod condition for bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/test-pod to (PodScheduled==False, Reason=SchedulerError)
E0622 14:00:14.250557  110366 scheduler.go:596] error binding pod: scheduler RunBindPlugins failed for pod bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/test-pod: code 1, err <nil>
I0622 14:00:14.252858  110366 wrap.go:47] GET /api/v1/namespaces/bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/pods/test-pod: (1.679811ms) 200 [scheduler.test/v0.0.0 (linux/amd64) kubernetes/$Format 127.0.0.1:37400]
I0622 14:00:14.253692  110366 reflector.go:243] k8s.io/client-go/informers/factory.go:133: forcing resync
I0622 14:00:15.247011  110366 reflector.go:243] k8s.io/client-go/informers/factory.go:133: forcing resync
I0622 14:00:15.247064  110366 reflector.go:243] k8s.io/client-go/informers/factory.go:133: forcing resync
I0622 14:00:15.247078  110366 reflector.go:243] k8s.io/client-go/informers/factory.go:133: forcing resync
I0622 14:00:15.247118  110366 reflector.go:243] k8s.io/client-go/informers/factory.go:133: forcing resync
```

Reply:
The reason the scheduler tries to schedule the pod multiple times is that the scheduler is running on a different thread, and could have tried to schedule the pod again before another poll is triggered.

Another problem we have in this test file in general is that we rely on shared state across tests (the counters), and each test relies on the other to reset that state, we should refactor this whole file to make sure we enforce that the state is reset before each test is run.

Reply:
I opened a PR https://github.com/kubernetes/kubernetes/pull/79424 to add reset methods to all test plugins and reset them after each test.

Reply:
Even if the scheduler retries scheduling a pod and determines that the pod is unschedulable, it shouldn't call "Unreserve" plugins. Unreserve plugins are called only if a pod passes the "reserve" point and later on is rejected and not get bound. The problem is probably shared state that @ahg-g referred to.

Reply:
> Even if the scheduler retries scheduling a pod and determines that the pod is unschedulable, it shouldn't call "Unreserve" plugins. Unreserve plugins are called only if a pod passes the "reserve" point and later on is rejected and not get bound.

Why is that? I think "Unreserve" plugin would be called each time bind plugin returns an error, see https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/scheduler.go#L580-L593
And my previous comment has the log ```framework.go:274] bind plugin bind-plugin-1 failed to bind pod bind-plugin97b74d23-14d8-4525-9072-7fbfa7cbacad/test-pod: failed to bind``` showing multiple times which indicates the bind plugin did have been called multiple times.

While I still deleted my second commit to see if simply rebasing revert commit against master could solve my flaky test.

Reply:
> Why is that? I think "Unreserve" plugin would be called each time bind plugin returns an error, see https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/scheduler.go#L580-L593

That's one of the places that "unreserve" plugins are called. There are other places too:

https://github.com/kubernetes/kubernetes/blob/de8fb1c9f81b1bd48a0d25f866492acca2c199f0/pkg/scheduler/scheduler.go#L557

https://github.com/kubernetes/kubernetes/blob/de8fb1c9f81b1bd48a0d25f866492acca2c199f0/pkg/scheduler/scheduler.go#L576

That said, I see your point about bind being called multiple times. In that case, we should expect unreserve to be called the same number of times are bind is called. Due to race conditions it could sometimes be called 'bind - 1' number of times if we check "after bind, and before unreserve".